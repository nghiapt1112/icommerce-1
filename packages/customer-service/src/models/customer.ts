import { Sequelize, Model, DataTypes, InferAttributes, InferCreationAttributes, CreationOptional } from 'sequelize';

class Customer extends Model<InferAttributes<Customer>, InferCreationAttributes<Customer>> {
  declare id: CreationOptional<string>;

  declare firstName: string;

  declare lastName: string;

  declare phone: string;

  declare createdAt: CreationOptional<Date>;

  declare updatedAt: CreationOptional<Date>;
}

export const defineOrder = (sequelize: Sequelize, model: typeof Customer) => {
  model.init(
    {
      id: {
        type: DataTypes.UUID,
        defaultValue: DataTypes.UUIDV4,
        primaryKey: true,
        allowNull: false,
      },
      firstName: {
        type: DataTypes.STRING,
        allowNull: false,
      },
      lastName: {
        type: DataTypes.STRING,
        allowNull: false,
      },
      phone: {
        type: DataTypes.STRING,
        allowNull: false,
      },
      createdAt: {
        type: DataTypes.DATE,
      },
      updatedAt: {
        type: DataTypes.DATE,
      },
    },
    {
      sequelize,
      modelName: 'Customer',
      tableName: 'customers',
      timestamps: true,
    },
  );
};

export default Customer;
