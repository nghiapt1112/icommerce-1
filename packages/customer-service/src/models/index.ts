import { dbConnection } from '../config';
import Customer, { defineOrder } from './customer';

export const defineModels = (conn = dbConnection) => {
  defineOrder(conn, Customer);
};

export { Customer };
