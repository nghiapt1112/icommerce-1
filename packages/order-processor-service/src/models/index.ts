import { dbConnection } from '../config';
import { Order, defineOrder } from './order';
import { OrderItem, defineOrderItem } from './orderitem';
import { Transaction, defineTransaction } from './transaction';

export const defineModels = (conn = dbConnection) => {
  defineOrder(conn, Order);
  defineOrderItem(conn, OrderItem);
  defineTransaction(conn, Transaction);

  Order.hasMany(OrderItem, { foreignKey: 'orderId', as: 'orderItems' });
  Order.hasOne(Transaction, { foreignKey: 'orderId' });
};

export { Order, OrderItem, Transaction };
