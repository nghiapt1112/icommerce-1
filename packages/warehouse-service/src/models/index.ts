import { dbConnection } from '../config';
import Product, { defineProduct } from './product';

export const defineModels = (conn = dbConnection) => {
  defineProduct(conn, Product);
};

export { Product };
