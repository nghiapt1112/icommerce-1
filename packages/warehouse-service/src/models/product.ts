import { Sequelize, Model, DataTypes, InferAttributes, InferCreationAttributes, CreationOptional } from 'sequelize';

class Product extends Model<InferAttributes<Product>, InferCreationAttributes<Product>> {
  declare id: CreationOptional<string>;

  declare name: string;

  declare price: number;

  declare branch: string;

  declare color: string;

  declare sku: number;

  declare createdAt: CreationOptional<Date>;

  declare updatedAt: CreationOptional<Date>;
}

export const defineProduct = (sequelize: Sequelize, model: typeof Product) => {
  model.init(
    {
      id: {
        type: DataTypes.UUID,
        defaultValue: DataTypes.UUIDV4,
        primaryKey: true,
        allowNull: false,
      },
      name: {
        type: DataTypes.STRING,
        allowNull: false,
      },
      price: {
        type: DataTypes.DECIMAL(10, 2),
        allowNull: false,
      },
      branch: {
        type: DataTypes.STRING,
        allowNull: false,
      },
      color: {
        type: DataTypes.STRING,
        allowNull: false,
      },
      sku: {
        type: DataTypes.INTEGER,
        defaultValue: 0,
      },
      createdAt: {
        type: DataTypes.DATE,
      },
      updatedAt: {
        type: DataTypes.DATE,
      },
    },
    {
      sequelize,
      modelName: 'Product',
      tableName: 'products',
      timestamps: true,
      indexes: [
        {
          fields: ['branch', 'color'],
        },
      ],
    },
  );
};

export default Product;
