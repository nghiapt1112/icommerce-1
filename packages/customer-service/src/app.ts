import config from 'config';
import { Microservice } from '@lib/microservice';
import { logger, dbConnection } from './config';
import { defineModels } from './models';
import { getCustomerById, createCustomer } from './routes';

const { app, listen } = Microservice({
  serviceName: config.get('serviceName'),
  isProduction: process.env.NODE_ENV === 'production',
  serverCors: false,
  jwtConfigurations: {
    publicKey: config.get('server.jwt.publicKey'),
    privateKey: config.get('server.jwt.privateKey'),
    algorithm: config.get('server.jwt.algorithm'),
    expiresIn: config.get('server.jwt.expiresIn'),
    issuer: config.get('serviceName'),
    audience: config.get('server.jwt.audience'),
    audiencePublicKeys: config.get('server.jwt.audiencePublicKeys'),
    ignoredPaths: config.get('server.jwt.ignoredPaths'),
  },
  serverPort: config.get<number>('server.port'),
  serverJsonLimit: config.get('server.jsonLimit'),
  serverListenCb: async () => {
    defineModels();
  },
  serverExit: async () => {
    await dbConnection.close();
    logger.info('Database connection closed.');
  },
  logger,
});

app.post('/v1/customers', createCustomer);
app.get('/v1/customers/:id', getCustomerById);

export { app, listen };
