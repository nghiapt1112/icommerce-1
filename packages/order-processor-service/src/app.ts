import config from 'config';
import { Microservice } from '@lib/microservice';
import { logger, dbConnection } from './config';
import { createOrder } from './routes';
import { defineModels } from './models';

const { app, listen } = Microservice({
  serviceName: config.get('serviceName'),
  isProduction: process.env.NODE_ENV === 'production',
  serverCors: false,
  jwtConfigurations: {
    secret: config.get('server.jwt.secret'),
    expiresIn: config.get('server.jwt.expiresIn'),
    issuer: config.get('serviceName'),
    audience: config.get('server.jwt.audience'),
    ignoredPaths: config.get('server.jwt.ignoredPaths'),
  },
  serverPort: config.get<number>('server.port'),
  serverJsonLimit: config.get('server.jsonLimit'),
  serverListenCb: async () => {
    defineModels();
  },
  serverExit: async () => {
    await dbConnection.close();
    logger.info('Database connection closed.');
  },
  logger,
});

app.post('/v1/orders', createOrder);

export { app, listen };
