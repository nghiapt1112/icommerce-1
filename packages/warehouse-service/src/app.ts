import config from 'config';
import { Microservice } from '@lib/microservice';
import { logger, dbConnection } from './config';
import { getProducts, getProductById } from './routes';
import { defineModels } from './models';

const { app, listen } = Microservice({
  serviceName: config.get('serviceName'),
  isProduction: process.env.NODE_ENV === 'production',
  serverCors: false,
  serverApiKey: config.get('server.apiKey'),
  serverPort: config.get<number>('server.port'),
  serverJsonLimit: config.get('server.jsonLimit'),
  // eslint-disable-next-line @typescript-eslint/require-await
  serverListenCb: async () => {
    defineModels();
  },
  serverExit: async () => {
    await dbConnection.close();
    logger.info('Database connection closed.');
  },
  logger,
});

app.get('/v1/products', getProducts);
app.get('/v1/products/:id', getProductById);

export { app, listen };
