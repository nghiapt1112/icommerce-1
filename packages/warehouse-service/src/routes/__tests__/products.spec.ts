/* eslint-disable @typescript-eslint/no-unsafe-argument */
/* eslint-disable @typescript-eslint/no-unsafe-return */

import _get from 'lodash/get';
import request, { Response } from 'supertest';
import { app } from '../../app';
import { Product } from '../../models';
import { productFactory } from '../../test';

const mockServiceGetById = jest.fn();
const mockServiceGetAll = jest.fn();
jest.mock('../../services/product.service', () => ({
  getById: (...args: any) => mockServiceGetById(...args),
  getAll: (...args: any) => mockServiceGetAll(...args),
}));

jest.mock('@lib/common', () => ({
  ...jest.requireActual('@lib/common'),
  genAnonymousUserIdFromRequest: jest.fn().mockReturnValue('test-id'),
}));

const mockRabbitMqPublish = jest.fn();
const mockRabbitMqConstructEvent = jest.fn();
jest.mock('@lib/server', () => ({
  ...jest.requireActual('@lib/server'),
  rabbitmq: {
    publish: (...args: any) => mockRabbitMqPublish(...args),
    constructEvent: (...args: any) => mockRabbitMqConstructEvent(...args),
  },
}));

jest.useFakeTimers().setSystemTime(new Date('2022-01-01').getTime());

describe('Product Endpoints', () => {
  afterEach(() => {
    mockServiceGetById.mockReset();
    mockServiceGetAll.mockReset();
  });

  test('error: GET /v1/products/:id', async () => {
    const productId = '44777823-4664-452d-baf3-2d207f0ce3e7';
    mockServiceGetById.mockImplementation(() => {
      throw new Error('product not found');
    });

    const response: Response = await request(app)
      .get(`/v1/products/${productId}`)
      .set('Accept', 'application/json')
      .send();

    expect(response.statusCode).toBe(404);
    expect(response.body).toEqual({ message: 'product not found' });
  });

  test('succeeded: GET /v1/products/:id', async () => {
    const productId = 'f2035fd2-22a8-4827-bc8e-a6896d213885';
    mockServiceGetById.mockImplementation(() =>
      Product.build(
        productFactory.build({
          id: productId,
          name: 'test',
          price: 2.99,
          branch: 'luxury',
          color: 'red',
          sku: 10,
          createdAt: new Date(),
          updatedAt: new Date(),
        }),
      ),
    );

    mockRabbitMqConstructEvent.mockImplementation(() => ({ name: 'test' }));

    const response: Response = await request(app)
      .get(`/v1/products/${productId}`)
      .set('Accept', 'application/json')
      .send();

    expect(response.statusCode).toBe(200);
    expect(_get(response.body, 'id')).toBe(productId);

    expect(mockRabbitMqConstructEvent).toHaveBeenCalledTimes(1);
    expect(mockRabbitMqConstructEvent.mock.calls[0]).toMatchSnapshot();

    expect(mockRabbitMqPublish).toHaveBeenCalledTimes(1);
    expect(mockRabbitMqPublish.mock.calls[0]).toMatchSnapshot();
  });

  test('GET /v1/products', async () => {
    mockServiceGetAll.mockImplementation(() => ({
      products: [
        Product.build(
          productFactory.build({
            id: '35648c7d-1819-4556-a0ce-c7e496089880',
            name: 'test',
            price: 2.99,
            branch: 'luxury',
            color: 'red',
            sku: 10,
            createdAt: new Date(),
            updatedAt: new Date(),
          }),
        ),
      ],
    }));

    const response: Response = await request(app).get(`/v1/products`).set('Accept', 'application/json').send();

    expect(response.statusCode).toBe(200);

    expect(mockRabbitMqConstructEvent).toHaveBeenCalledTimes(1);
    expect(mockRabbitMqConstructEvent.mock.calls[0]).toMatchSnapshot();

    expect(mockRabbitMqPublish).toHaveBeenCalledTimes(1);
    expect(mockRabbitMqPublish.mock.calls[0]).toMatchSnapshot();
  });
});
