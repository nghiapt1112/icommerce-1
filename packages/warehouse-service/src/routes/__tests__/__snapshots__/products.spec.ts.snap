// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Endpoints GET /v1/products 1`] = `
Array [
  "user:product:searching:filtering",
  "warehouse-service:products",
  Object {
    "matchedProductIds": Array [
      "35648c7d-1819-4556-a0ce-c7e496089880",
    ],
    "query": Object {},
  },
  Object {
    "requestinfo": undefined,
    "userid": undefined,
  },
]
`;

exports[`Product Endpoints GET /v1/products 2`] = `
Array [
  undefined,
  "user:activities",
  Object {
    "name": "test",
  },
  Object {
    "error": [Function],
    "info": [Function],
    "warn": [Function],
  },
]
`;

exports[`Product Endpoints succeeded: GET /v1/products/:id 1`] = `
Array [
  "user:product:viewed",
  "warehouse-service:product:f2035fd2-22a8-4827-bc8e-a6896d213885",
  Object {
    "branch": "luxury",
    "color": "red",
    "createdAt": 2022-01-01T00:00:00.000Z,
    "id": "f2035fd2-22a8-4827-bc8e-a6896d213885",
    "name": "test",
    "price": 2.99,
    "sku": 10,
    "updatedAt": 2022-01-01T00:00:00.000Z,
  },
  Object {
    "requestinfo": undefined,
    "userid": undefined,
  },
]
`;

exports[`Product Endpoints succeeded: GET /v1/products/:id 2`] = `
Array [
  undefined,
  "user:activities",
  Object {
    "name": "test",
  },
  Object {
    "error": [Function],
    "info": [Function],
    "warn": [Function],
  },
]
`;
